### PR TITLE
[SPARK-24836][SQL] New option for Avro datasource - ignoreExtension

### PR DESCRIPTION
**Pull request by MaxGekk**
## What changes were proposed in this pull request?

I propose to add new option for AVRO datasource which should control ignoring of files without `.avro` extension in read. The option has name `ignoreExtension` with default value `true`. If both options `ignoreExtension` and `avro.mapred.ignore.inputs.without.extension` are set, `ignoreExtension` overrides the former one. Here is an example of usage:

```
spark
  .read
  .option("ignoreExtension", false)
  .avro("path to avro files")
```

## How was this patch tested?

I added a test which checks the option directly and a test for checking that new option overrides hadoop's config.

### CODE REVIEW

File: external/avro/src/main/scala/org/apache/spark/sql/avro/AvroFileFormat.scala [hunk 1]
```diff
@@ -276,10 +274,15 @@ private[avro] object AvroFileFormat {
     }
   }
 
-  def ignoreFilesWithoutExtensions(conf: Configuration): Boolean = {
-    // Files without .avro extensions are not ignored by default
-    val defaultValue = false
+  def ignoreExtension(conf: Configuration, options: Map[String, String]): Boolean = {
```
Comment (gatorsmile) [inline external/avro/src/main/scala/org/apache/spark/sql/avro/AvroFileFormat.scala]: Could we have a class `AvroOptions` like what we are doing for the other built-in data sources?

Comment (MaxGekk) [reply external/avro/src/main/scala/org/apache/spark/sql/avro/AvroFileFormat.scala]: Would you like to see it as a part of this PR or a separate one? I would extract some common code like `getBool()` from CSVOptions to a separate trait and extend `AvroOptions` by it.

File: external/avro/src/main/scala/org/apache/spark/sql/avro/AvroFileFormat.scala [hunk 2]
```diff
@@ -276,10 +274,15 @@ private[avro] object AvroFileFormat {
...
+    val ignoreFilesWithoutExtensionByDefault = false
+    val ignoreFilesWithoutExtension = conf.getBoolean(
+      AvroFileFormat.IgnoreFilesWithoutExtensionProperty,
+      ignoreFilesWithoutExtensionByDefault)
 
-    conf.getBoolean(AvroFileFormat.IgnoreFilesWithoutExtensionProperty, defaultValue)
+    options
+      .get("ignoreExtension")
```
Comment (viirya) [inline external/avro/src/main/scala/org/apache/spark/sql/avro/AvroFileFormat.scala]: Shall we document this option somewhere?

Comment (MaxGekk) [reply external/avro/src/main/scala/org/apache/spark/sql/avro/AvroFileFormat.scala]: Sure, I am going to add the `AvroOption` class and document all Avro options there.

Comment (HyukjinKwon) [reply external/avro/src/main/scala/org/apache/spark/sql/avro/AvroFileFormat.scala]: Let's make sure we describe that in a public API later.

**Comment by MaxGekk** [comment]
Please, look at this PR: https://github.com/apache/spark/pull/21810 . It introduces `AvroOptions`.

**Comment by MaxGekk** [comment]
I added new option to `AvroOptions` and documented it. Please, look at the PR one more time.

File: external/avro/src/main/scala/org/apache/spark/sql/avro/AvroFileFormat.scala [hunk 3]
```diff
@@ -275,10 +273,12 @@ private[avro] object AvroFileFormat {
     }
   }
 
-  def ignoreFilesWithoutExtensions(conf: Configuration): Boolean = {
-    // Files without .avro extensions are not ignored by default
-    val defaultValue = false
+  def ignoreExtension(conf: Configuration, options: AvroOptions): Boolean = {
```
Comment (gatorsmile) [inline external/avro/src/main/scala/org/apache/spark/sql/avro/AvroFileFormat.scala]: Can we move this to `object AvroOptions`?

Comment (gatorsmile) [reply external/avro/src/main/scala/org/apache/spark/sql/avro/AvroFileFormat.scala]: Also add a comment above this function to describe how we determine it?

**Comment by gatorsmile** [comment]
LGTM except a comment

**Comment by gatorsmile** [comment]
LGTM

Thanks! Merged to master.

**Pull request closed without merging**